test/should_pass/operator_subtypes.erl

```diff
@@ -104,3 +104,18 @@ unary_op4(X) -> -X.
 unary_op5(X) ->
     A = bnot X,
     A.
+
+-spec unary_op6(pos_integer()) -> neg_integer().
+unary_op6(X) ->
+    A = -X,
+    A.
```

Comment:
Why not just `unary_op6(X) -> -X.`? Does that cause code coverage to be different?

Reply:
Yes. It happens a fair bit due to bidirectionality. I think the code you suggested is already covered elsewhere. If it's not, I'll find it in another coverage hunt!

Reply:
An alternative to using the trick with variables is to use `?annotate_type(Expr, any())`. It's also a bit obscure. Perhaps it's worth defining a new macro with a helpful name to use in these types of test where we want to make sure we test inference (inference, in the sense that we don't know what the final type should be, not in the `--infer` sense) rather than checking.

Reply:
A new macro to be used in the tests is a great idea! Can you show how `?annotate_type(Expr, any())` would be used here to verify that `-X :: neg_integer()`? Can the new macro be crafted to verify the exact type instead of just compatibility?

Reply:
Hmmm, `?annotate_type` wouldn't help to verify `-X :: neg_integer()`. 

I feel a little uneasy about the `X = <bla>, X.` pattern because ideally we would like gradualizer to actually typecheck this better. In the future I would like the typechecking of variable bindings to go like this: create a type variable, let's call it `T` and give `X` the type `T`. Then typecheck `<bla>` with the expected type `T`. In the rest of the expression, use `X :: T` in the environment. This will propagate types better that the way we're doing it now, but relies on having a constraint solver in place. 

I think a more robust syntax would be to say `case <bla> of X -> X.` That could be made into a macro.

Reply:
Another idea would be to unfold the variables, effectively rewriting the function to `unary_op6(X) -> -X`. To do this, we can bind the variable to the expression instead of a type. Then we have `A = -X` in the environment and whenever we run into `type_check_expr_ty(A,  neg_integer())` it expands to `type_check_expr_ty(-X, neg_integer())` (pseudo-code).

I'm trying to avoid relying on constraint-solving because, after reading the paper POPL19 which covers polymorphism with set-theoretic gradual types, it seems very hard. They solve it (by completely redefining stuff, replacing the compatibility relation with a "materialization" rule, "frame variables", etc etc), but then instead, intersection types for function types are not supported in that system. It's listed under *future work*:

*"This work lays a foundation for integrating gradual typing and full set-theoretic types and, as such,
it opens many new questions and issues. There are in particular two practical issues that we want
to address in the near future.*

*The first is to address a restriction we imposed to our system: namely, that it is not possible
to assign intersection types to a function. Forbidding that (other than by subsumption) was an
early design choice of this work, motivated by several reasons: its absence would complicate the
dynamic semantics of the cast calculus (see Castagna and Lanvin [2017], where this restriction is
not present); it would make type reconstruction and constraint solving much more difficult, and
it would have probably hindered completeness even for simple systems; a system without this
restriction would have been interesting only if the language had a type-case construct, which we
wanted to avoid for simplicity and for sticking as close as possible to ML."*